Log a verbose message when DOTNET_CLI_HOME is being used.

peterhuene opened the pull request:
This commit logs a diagnostic message when the `DOTNET_CLI_HOME` variable is
used.  This enables users to determine where first-run-experience and global
tool files are being written to.

Fixes #9510.

peterhuene commented:
Ping @nkolev92.  I opted for a "verbose" message regarding the `DOTNET_CLI_HOME` variable being used because we have commands like `dotnet run` that should not display output by default.  With the `-d` option passed to `dotnet`, this will at least enable users to determine if their home directory is being overridden.  Let me know what you think since you are the one who requested this behavior.  Thanks!

nkolev92 commented:
I think that's reasonable!

Thanks!

The biggest reason for this is the fact that the global packages folder and the plugins directory change.
If/when customers see some weird issues, we can ask them to do verbose logs and it'll indicate clearly what's going on.

wli3 commented:
I assume you talked to Kathleen

wli3 commented on src/dotnet/Program.cs:
```diff
@@ -149,6 +149,8 @@ internal static int ProcessArgs(string[] args, ITelemetry telemetryClient = null
                         bool skipFirstRunExperience =
                             environmentProvider.GetEnvironmentVariableAsBool("DOTNET_SKIP_FIRST_TIME_EXPERIENCE", false);
 
+                        ReportDotnetHomeUsage(environmentProvider);
```
nit: why here? not somewhere close to Reporter.Output.WriteLine("Performance Summary:");? Or else?

It's better to avoid adding stuff to Progam.cs but I don't know a better place

peterhuene replied:
I wanted the output to be before command execution, close to where we would report telemetry being used or not, simply because this variable can alter the behavior of the command that's about to run.  It also affects first-run experience, so I'd like it printed before that too.

peterhuene commented:
I actually haven't, so ping @KathleenDollard 😄 

In a previous change into 2.1.4xx, we fixed the crash / unexpected behavior of when `HOME` (macOS/Linux) or `USERPROFILE` (Windows) was not set for the dotnet process.  As part of that change, I added support for a new variable called `DOTNET_CLI_HOME` which dotnet would respect as those variables if set.  However, NuGet still required the home environment variables to be set to work at the time.  This has now been addressed such that NuGet will work with just `DOTNET_CLI_HOME` set too, enabling CI scenarios that want to instruct the .NET Core SDK (and, by extension, NuGet) to put its files for users without a home directory.

It was requested that the CLI log a message when this variable is being preferred so that users know where the files will be put.  This commit fulfills that request, although only at the "diagnostic" output level with `dotnet -d`.

peterhuene commented:
@Pilchie ping for approval.

Pilchie commented:
Approved for P5, but I wonder about the loc constraints here.  Will we be able to merge this in time to get a loc handback @nguerrera ?

nguerrera commented:
Hold merge until we decide we're open for P5.

> I wonder about the loc constraints here.

@Pilchie We need to have all string changes merged into release/2.1.4xx *before* July 9th.

If we're holding checkins for P5 until after that, then we have a problem. It is possible that we can negotiate something else (use a different branch, arrange a one off loc cycle, etc), but I'd really prefer if we can stick to the plan with everything else going on.

Can we set a time before July 9th after which we're considered done for P4 and then merge these?

peterhuene commented:
Marking WIP until final approval for P5.
